File: epms/employee_performance/report/low_performers/low_performers.py
import frappe
from frappe import _
from frappe.utils import getdate, nowdate, cint


def execute(filters=None):
    """Execute Low Performers Report."""
    if not filters:
        filters = {}

    columns = get_columns()
    data = get_data(filters)
    chart = get_chart_data(data)

    return columns, data, None, chart


def get_columns():
    """Get report columns."""
    return [
        {
            "fieldname": "employee_name",
            "fieldtype": "Data",
            "label": _("Employee"),
            "width": 150,
        },
        {
            "fieldname": "team",
            "fieldtype": "Link",
            "label": _("Team"),
            "options": "Team",
            "width": 120,
        },
        {
            "fieldname": "overall_score",
            "fieldtype": "Float",
            "label": _("Overall Score"),
            "width": 110,
        },
        {
            "fieldname": "final_grade",
            "fieldtype": "Data",
            "label": _("Grade"),
            "width": 100,
        },
        {
            "fieldname": "productivity_score",
            "fieldtype": "Float",
            "label": _("Productivity"),
            "width": 100,
        },
        {
            "fieldname": "quality_score",
            "fieldtype": "Float",
            "label": _("Quality"),
            "width": 80,
        },
        {
            "fieldname": "attendance_score",
            "fieldtype": "Float",
            "label": _("Attendance"),
            "width": 90,
        },
        {
            "fieldname": "tasks_completed",
            "fieldtype": "Int",
            "label": _("Tasks Completed"),
            "width": 110,
        },
        {
            "fieldname": "pending_tasks",
            "fieldtype": "Int",
            "label": _("Pending Tasks"),
            "width": 100,
        },
        {
            "fieldname": "performance_status",
            "fieldtype": "Data",
            "label": _("Status"),
            "width": 100,
        },
    ]


def get_data(filters):
    """Get report data."""
    month = cint(filters.get("month", getdate(nowdate()).month))
    year = cint(filters.get("year", getdate(nowdate()).year))
    threshold = cint(filters.get("threshold", 60))

    conditions = {
        "month": month,
        "year": year,
        "overall_score": ["<", threshold],
        "docstatus": 1,
    }

    if filters.get("team"):
        conditions["team"] = filters["team"]

    data = frappe.get_all(
        "Performance Scorecard",
        filters=conditions,
        fields=[
            "employee_name",
            "team",
            "overall_score",
            "final_grade",
            "productivity_score",
            "quality_score",
            "attendance_score",
            "tasks_completed",
            "pending_tasks",
            "performance_status",
        ],
        order_by="overall_score asc",
    )

    return data


def get_chart_data(data):
    """Get chart data for the report."""
    if not data:
        return None

    return {
        "data": {
            "labels": [d.get("employee_name", "Unknown") for d in data],
            "datasets": [
                {
                    "name": "Overall Score",
                    "values": [d.get("overall_score", 0) for d in data],
                }
            ],
        },
        "type": "bar",
        "colors": ["#dc3545"],
    }

File: epms/employee_performance/report/low_performers/low_performers.js
function epms_badge(v){if(!v)return'';var s=String(v).toLowerCase(),c='epms-badge-info';if(['on track','excellent','very good','completed','a+','a','b+','b','gold','silver'].indexOf(s)>-1)c='epms-badge-success';else if(['needs attention','good','average','in progress','c','d'].indexOf(s)>-1)c='epms-badge-warning';else if(['at risk','needs improvement','blocked','f','fail'].indexOf(s)>-1)c='epms-badge-danger';else if(['pending'].indexOf(s)>-1)c='epms-badge-muted';return'<span class="epms-badge '+c+'">'+v+'</span>';}
function epms_score(v){if(v===null||v===undefined||v==='')return'0';var n=parseFloat(v)||0,c='epms-badge-success';if(n<60)c='epms-badge-danger';else if(n<80)c='epms-badge-warning';return'<span class="epms-badge '+c+'">'+n.toFixed(1)+'</span>';}
function epms_fmt(spec){setTimeout(function(){var t=document.querySelector('.page-wrapper table');if(!t)return;var hs=[];t.querySelectorAll('thead th').forEach(function(th){hs.push(th.textContent.trim().toLowerCase());});t.querySelectorAll('tbody tr').forEach(function(r){var cs=r.querySelectorAll('td');hs.forEach(function(h,i){if(spec[h]&&cs[i])cs[i].innerHTML=spec[h](cs[i].textContent.trim());});});},1500);}

frappe.query_reports['Low Performers']={
    filters:[
        {fieldname:'month',fieldtype:'Select',label:__('Month'),options:'1\n2\n3\n4\n5\n6\n7\n8\n9\n10\n11\n12',default:new Date().getMonth()+1,reqd:1},
        {fieldname:'year',fieldtype:'Int',label:__('Year'),default:new Date().getFullYear(),reqd:1},
        {fieldname:'team',fieldtype:'Link',label:__('Team'),options:'Team'},
        {fieldname:'threshold',fieldtype:'Int',label:__('Score Threshold'),default:60,description:__('Show employees with score below this value')}
    ],
    onload:function(r){epms_fmt({'overall score':epms_score,'grade':epms_badge,'status':epms_badge});}
};
